src/pybullet_industrial/g_code_processor.py

```diff
@@ -0,0 +1,405 @@
+                self.execute_t_cmd(cmd)
+
+            elif cmd_type == "G":
+                self.excecute_g_cmd(cmd)
+
+    def excecute_m_cmd(self, cmd):
+        self.m_commands[cmd[0][1]][0]()
+
+    def execute_t_cmd(self, cmd):
+        self.t_commands[cmd[0][1]][0]()
+        self.calibrate_tool()
+
+    def excecute_g_cmd(self, cmd):
+
+        self.last_point = np.array(self.new_point)
+        self.last_or = np.array(self.new_or)
+
+        g_cmd_type = cmd[0][1]
+
+        # Activation of the zero offset
+        if g_cmd_type == 54:
+            self.offset = np.array([self.last_point, self.last_or])
+
+        # Deactivation of the zero offset
+        elif g_cmd_type == 500:
+            self.offset = np.array([[0.0, 0.0, 0.0],
+                                    [0.0, 0.0, 0.0]])
+
+        # Axis selelection for circular interpolation
+        elif g_cmd_type == 17:
+            self.axis = 2  # X-Y Axis
+
+        elif g_cmd_type == 18:
+            self.axis = 1  # X-Z Axis
+
+        elif g_cmd_type == 19:
+            self.axis = 0  # Y-Z Axis
+
+        elif g_cmd_type in [0, 1, 2, 3]:
+            variables = {'X': np.nan, 'Y': np.nan, 'Z': np.nan, 'A': np.nan,
+                         'B': np.nan, 'C': np.nan, 'R': np.nan}
+
+            for val in cmd:
+                if val[0] in variables:
+                    variables[val[0]] = val[1]
+
+            xyz_val = np.array([variables['X'], variables['Y'],
+                                variables['Z']])
+            abc_val = np.array([variables['A'], variables['B'],
+                                variables['C']])
+            r_val = variables['R']
+
+            # Setting the new point considering offset
+            self.new_point = np.array([0.0, 0.0, 0.0])
+            for i, value in enumerate(xyz_val):
+                if np.isnan(value):
+                    self.new_point[i] = self.last_point[i]
+                else:
+                    self.new_point[i] = value + self.offset[0][i]
+
+            # Setting the new orientation considering offset
+            self.new_or = np.array([0.0, 0.0, 0.0])
+            for i, value in enumerate(abc_val):
+                if np.isnan(value):
+                    self.new_or[i] = self.last_or[i]
+                else:
+                    self.new_or[i] = value + self.offset[1][i]
+
+            # Moving the endeffector if there is a G0 Interpolation
+            if g_cmd_type == 0:
+                self.g0_interpolation()
+
+            # Moving the endeffector if ther is a G1 Interpolation
+            elif g_cmd_type in [1, 2, 3]:
+                self.g123_interpolation(g_cmd_type, r_val)
+
+    def g0_interpolation(self):
+        """The G0-interpolation sets the position of the endeffector
+        or tool without running an interpolation command
+
+        Args:
+            None
+        Returns:
+            None
+        """
+        orientation = p.getQuaternionFromEuler(self.new_or)
```

Comment:
Comment on old gcode_class: "Make sure you check that the abc euler conversion of G-code corresponds to the Eular angle definition of pybullet. Note that there are several different euler angle defitions.

@liquidcronos Could you help me with this? Shouldn't the euler conversion of the pybullet always follow the default value?

This probably warants a test case"

Reply:
There is not really a default euler conversion. Different Fields define different euler angles around different axis.
Check the [wikipedia article](https://en.wikipedia.org/wiki/Euler_angles) for reference.

I believe that pybullet uses the RPY (Roll Pitch Yaw) formulation. This means that the first vector describes a rotation around the x axis, the second around the y axis and the third around the z axis. We need to make sure that the deafult g code formulation of a, b, c corresponds to the same euler angles and not a different convention.